rlclientlib/federation/sender_joined_log_provider.cc

```diff
@@ -0,0 +1,244 @@
+    std::vector<flatbuffers::Offset<messages::flatbuff::v2::JoinedEvent>> events;
+
+    const auto& item = *it;
+    const auto interaction_timestamp = std::get<0>(it->first);
+    const auto interaction_time = interaction_timestamp.to_time_point();
+    if (interaction_time <= eud_cutoff)
+    {
+      const auto reward_cutoff = interaction_time + _eud_offset;
+
+      auto ts = from_rl_timestamp(std::get<0>(it->first));
+      events.push_back(messages::flatbuff::v2::CreateJoinedEventDirect(fbb, &it->second, &ts));
+      const auto& event_id = std::get<1>(it->first);
+      const auto& observations = _observations.find(event_id);
+      if (observations != _observations.end())
+      {
+        for (const auto& observation : observations->second)
+        {
+          const auto observation_timestamp = std::get<0>(observation);
+          const auto observation_time = observation_timestamp.to_time_point();
+          if (observation_time <= reward_cutoff)
+          {
+            auto ts = from_rl_timestamp(std::get<0>(observation));
+            events.push_back(messages::flatbuff::v2::CreateJoinedEventDirect(fbb, &std::get<1>(observation), &ts));
+          }
+        }
+        _observations.erase(event_id);
+      }
+
+      auto joined_payload = messages::flatbuff::v2::CreateJoinedPayloadDirect(fbb, &events);
+      emit_regular_message(result, fbb, joined_payload);
+      _interactions.erase(it++);
+    }
+    else
+    {
+      // TODO make sure this is correct. As in does the timestamp ordering mean that all further events are within eud.
+      // break;
+      ++it;
+    }
+  }
+
+  batch.reset(new buffer_reader(std::move(result)));
+
+  return 0;
+}
+
+int sender_joined_log_provider::receive_events(const i_sender::buffer& data, api_status* status)
+{
+  std::lock_guard<std::mutex> lock(_mutex);
+
+  logger::preamble pre;
+  pre.read_from_bytes(data->preamble_begin(), logger::preamble::size());
+
+  if (pre.msg_type != logger::message_type::fb_generic_event_collection)
+  {
+    RETURN_ERROR_LS(_trace_logger, status, invalid_argument)
+        << " Message type " << pre.msg_type << " cannot be handled.";
+  }
+
+  auto res = messages::flatbuff::v2::GetEventBatch(data->body_begin());
+  flatbuffers::Verifier verifier(data->body_begin(), data->body_filled_size());
+  auto result = res->Verify(verifier);
+
+  if (!result)
+  { RETURN_ERROR_LS(_trace_logger, status, invalid_argument) << "verify failed for fb_generic_event_collection"; }
+
+  if (res->metadata()->content_encoding()->str() != "IDENTITY")
+  { RETURN_ERROR_LS(_trace_logger, status, invalid_argument) << "Can only handle IDENTITY encoding"; }
+
+  for (auto serialized_event : *res->events())
+  {
+    const auto* serialized_payload = serialized_event->payload();
+    const auto* event = flatbuffers::GetRoot<messages::flatbuff::v2::Event>(serialized_payload->data());
+
+    std::string event_id = event->meta()->id()->str();
+    std::vector<uint8_t> joined_event_data(
```

Comment:
Is there a way to avoid mem copy?

Reply:
Perhaps a class the holds on to the shared_ptr<buffer> and provides helper methods to access data?